File: tests/fixtures/ecosystems/dart_comprehensive/classes.dart
/// Class patterns.

abstract class Shape {
  double get area;
  double get perimeter;

  String describe() => '${runtimeType}: area=$area';
}

class Circle extends Shape {
  final double radius;

  Circle(this.radius);

  @override
  double get area => 3.14159 * radius * radius;

  @override
  double get perimeter => 2 * 3.14159 * radius;
}

class Rectangle extends Shape {
  final double width;
  final double height;

  Rectangle(this.width, this.height);

  @override
  double get area => width * height;

  @override
  double get perimeter => 2 * (width + height);
}

class Person {
  final String name;
  final int age;

  Person(this.name, this.age);

  String greet() => 'Hi, I\'m $name';

  @override
  String toString() => 'Person($name, $age)';
}

class Employee extends Person {
  final String department;

  Employee(String name, int age, this.department) : super(name, age);

  @override
  String greet() => 'Hi, I\'m $name from $department';
}

File: tests/fixtures/ecosystems/typescript_comprehensive/classes.ts
/**
 * Class patterns for parser testing.
 */

abstract class Shape {
    abstract area(): number;
    abstract perimeter(): number;

    describe(): string {
        return `Shape with area ${this.area()}`;
    }
}

class Circle extends Shape {
    constructor(private radius: number) {
        super();
    }

    area(): number {
        return Math.PI * this.radius ** 2;
    }

    perimeter(): number {
        return 2 * Math.PI * this.radius;
    }
}

class Rectangle extends Shape {
    constructor(
        private width: number,
        private height: number,
    ) {
        super();
    }

    area(): number {
        return this.width * this.height;
    }

    perimeter(): number {
        return 2 * (this.width + this.height);
    }
}

class Animal {
    static count = 0;

    #name: string;
    protected species: string;

    constructor(name: string, species: string) {
        this.#name = name;
        this.species = species;
        Animal.count++;
    }

    get name(): string {
        return this.#name;
    }

    set name(value: string) {
        this.#name = value;
    }

    toString(): string {
        return `${this.#name} (${this.species})`;
    }
}

class Dog extends Animal {
    constructor(name: string) {
        super(name, "Canine");
    }

    bark(): string {
        return "Woof!";
    }
}

export { Shape, Circle, Rectangle, Animal, Dog };
